File: datastructures/maxheap/maxheap.cpp
#include <iostream>
#include <vector>

using namespace std;
#include "maxheap.hpp"


void MaxHeap::maxHeapify() {
    for (auto i = this->N / 2; i > 0; i--) {
        this->down(i);
    }
}

void MaxHeap::down(unsigned long k) {
    while (2*k <= this->N) {

        auto child = 2*k;

        // Set the child to the right leaf it is bigger
        if (2*k+1 <= this->N && this->arr[child] < this->arr[child+1]) {
            child += 1;
        }

        // if the parent is larger than the largest child, break
        if (this->arr[k] > this->arr[child]) break;

        // Otherwise, swap k with the largest child
        swap(this->arr[k], this->arr[child]);

        // the largest child idx now becomes k and w 'recurse'
        k = child;
    }

}

char MaxHeap::removeMax() {
    // Remove the top of the vector
    char c = this->arr[1];

    // Swap the bottom with the top
    swap(this->arr[this->N], this->arr[1]);

    // Remove the elem from th vector completely
    this->arr.pop_back();

    // decrement the size by 1
    this->N -= 1;

    // Sink down the first element
    this->down(1);

    return c;
}

void MaxHeap::insert(char c) {
    this->arr.push_back(c);
    this->N += 1;
    this->up(this->N);
}

void MaxHeap::up(unsigned long k) {

    // While we aren't at 1 and the parent is greater than the child
    while (k >= 1 and this->arr[k/2] < this->arr[k]) {

        // swap the parent and child valus so the parent is now greater
        swap(this->arr[k/2], this->arr[k]);

        // set k to the parent value.
        k = k / 2;
    }
}

void MaxHeap::print() {
    for (auto i = 1; i <= N; i++) {
        cout << this->arr[i] << " ";
    }
    cout << endl;
}

int main() {
    MaxHeap maxHeap;
    maxHeap.insert('P');
    maxHeap.insert('Q');
    maxHeap.insert('E');
    maxHeap.print();
    cout << "removeMax: " << maxHeap.removeMax() << endl;
    maxHeap.print();
    maxHeap.insert('X');
    maxHeap.insert('A');
    maxHeap.insert('M');
    maxHeap.print();
    cout << "removeMax: " << maxHeap.removeMax() << endl;
    maxHeap.print();
    maxHeap.insert('P');
    maxHeap.insert('L');
    maxHeap.insert('E');
    maxHeap.print();
    cout << "removeMax:" << maxHeap.removeMax() << endl;
    maxHeap.print();

    // Final ordr will be P, M, L, A, E, E
    return 0;
}

File: datastructures/maxheap/maxheap.py
from __future__ import print_function


class MaxHeap:

    def __init__(self, arr, N):
        self._arr = arr
        self.N = N
        for i in range(N/2, 0, -1):
            self.down(i)

    def up(self, k):
        # While we aren't at the root and the parent is less than the current node, swap them
        while k > 0 and self._arr[k/2] < self._arr[k]:
            # Swap k/2 and k, i.e swap the parent with the current node
            self._arr[k/2], self._arr[k] = self._arr[k], self._arr[k/2]
            # Now go up to the parent and re-evaluate
            k /= 2

    # O(N)
    def down(self, k):
        # While we aren't at the last leaf.  2*k is the left child
        while 2*k <= self.N:
            # Left child of k
            j = 2*k

            # Choose the greater of the two child nodes
            if j+1 <= self.N and self._arr[j] < self._arr[j+1]:
                j += 1

            # If the parent greater than the greatest child, break, nothing more to do here
            if self._arr[k] > self._arr[j]:
                break

            # Exchange the parent and child
            self._arr[k], self._arr[j] = self._arr[j], self._arr[k]

            # The child is the new parent, and we have another go
            # I.e. whatever child was exchanged with the parent is the new parent
            k = j

    # O(N lg N)
    def sort(self):
        while self.N >= 1:

            yield self._arr[1]

            # Swap the first position with the last
            self._arr[1], self._arr[self.N] = self._arr[self.N], self._arr[1]

            # decrement the size
            self.N -= 1

            # Since just the top was swappd with the end, sink down the top now to the right place
            self.down(1)

            #  I don't believe running run max heapify fully here is necessary
            #for i in range(self.N/2, 0, -1):
            #    self.down(i)


if __name__ == '__main__':
    arr = [None, 1, 16, 4, 10, 4, 7, 9, 3, 2, 8, 1, 5]
    print(len(arr))
    print(arr)
    maxh = MaxHeap(arr, len(arr) - 1)
    print(maxh._arr)
    for i in maxh.sort():
        print(i, end=" ")
